httpclient5/src/main/java/org/apache/hc/client5/http/entity/LazyDecompressInputStream.java

```diff
@@ -0,0 +1,214 @@
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+package org.apache.hc.client5.http.entity;
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+
+import org.apache.commons.compress.compressors.CompressorException;
+import org.apache.hc.core5.io.Closer;
+
+
+/**
+ * A {@link FilterInputStream} that lazily initializes and applies decompression on the underlying input stream.
+ * This class supports multiple compression types and uses {@link CompressorFactory} to obtain the appropriate
+ * decompression stream when the first read operation occurs.
+ *
+ * <p>This implementation delays the creation of the decompression stream until it is required, optimizing
+ * the performance when the stream may not be read immediately or at all.</p>
+ *
+ * @since 5.5
+ */
+public class LazyDecompressInputStream extends FilterInputStream {
+
+    /**
+     * The lazily initialized decompression stream.
+     */
+    private InputStream wrapperStream;
+
+    /**
+     * The compression type used to determine which decompression algorithm to apply (e.g., "gzip", "deflate").
+     */
+    private final String compressionType;
+
+    /**
+     * The flag indicating if decompression should skip certain headers (noWrap).
+     */
+    private final boolean noWrap;
+
+    /**
+     * Constructs a new {@link LazyDecompressInputStream} that applies the specified compression type and noWrap setting.
+     *
+     * @param wrappedStream   the non-null {@link InputStream} to be wrapped and decompressed.
+     * @param compressionType the compression type (e.g., "gzip", "deflate").
+     * @param noWrap          whether to decompress without headers for certain compression formats.
+     */
+    public LazyDecompressInputStream(final InputStream wrappedStream, final String compressionType, final boolean noWrap) {
+        super(wrappedStream);
+        this.compressionType = compressionType;
+        this.noWrap = noWrap;
+    }
+
+    /**
+     * Constructs a new {@link LazyDecompressInputStream} that applies the specified compression type,
+     * defaulting to no noWrap handling.
+     *
+     * @param wrappedStream   the non-null {@link InputStream} to be wrapped and decompressed.
+     * @param compressionType the compression type (e.g., "gzip", "deflate").
+     */
+    public LazyDecompressInputStream(final InputStream wrappedStream, final String compressionType) {
+        this(wrappedStream, compressionType, false);
+    }
+
+    /**
+     * Initializes the decompression wrapper stream lazily, based on the compression type and noWrap flag.
+     *
+     * @return the initialized decompression stream.
+     * @throws IOException if an error occurs during initialization.
+     */
+    private InputStream initWrapper() throws IOException {
+        if (wrapperStream == null) {
```

Comment:
It does not look like there is any locking here, so the class should be Javadoc'd as thread unsafe.